<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES â€¢ IgniteRNAseq</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">IgniteRNAseq</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/FLAMESWorkflow.html">FLAMES package demo</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ChangqingW/IgniteRNAseq"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES</h1>
                        <h4 data-toc-skip class="author">Changqing
Wang</h4>
            <address class="author_afil">
      The Walter and Eliza Hall Institute of Medical
Research<br><h4 data-toc-skip class="date">16 July 2024</h4>
      

      <div class="d-none name"><code>FLAMESWorkflow.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>In this vignette, we will walk through the <code>FLAMES</code>
single-cell pipeline, from de-multiplexing raw FASTQ files to
identifying and quantifying transcripts. We then perform analysis
including single nucleotide polymorphism (SNP), differential transcript
expression analysis and differential transcript usage analysis. We will
use the cell line mixture data from Tian <em>et al.</em> <span class="citation">(<a href="#ref-tianComprehensiveCharacterizationSinglecell2021">Tian <em>et
al.</em>, 2021-11-11</a>)</span>.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>While the popular single-cell kits allows profiling the transcriptome
of up to 10,000 cells <span class="citation">(<a href="#ref-WhatMaximumNumber">What is the maximum number of cells that
can be profiled?</a>)</span> at once, the traditional short-read
sequencing (100-200 nts) limits our ability to unambiguously profile
<strong>isoform expression</strong> and <strong>mutations</strong>. To
overcome this issue, we can combine long-read sequencing with existing
single-cell kits.</p>
<p>Though long-read sequencing can provide read lengths several orders
of magnitude larger than the typical short-read sequencing, it comes at
much steeper cost and lower throughput, to accommodate this, we profile
a smaller sub-sample of cells in our <em>FLT-seq</em> protocol<span class="citation">(<a href="#ref-tianComprehensiveCharacterizationSinglecell2021">Tian <em>et
al.</em>, 2021-11-11</a>)</span>: <img src="FLT-seq.png" width="800"></p>
<p>There are also many other protocols that combine single-cell with
long-read sequencing, including <em>scRaCH-seq</em> which profiles gene
panel <span class="citation">(<a href="#ref-pengSinglecellRapidCapture2024">Peng <em>et al.</em>,
2024-01-30</a>)</span>, <em>LR-Split-seq</em> which uses combinatorial
barcoding and does not require microfluidic equipments <span class="citation">(<a href="#ref-rebboahMappingModelingGenomic2021">Rebboah <em>et al.</em>,
2021-10-07</a>)</span>.</p>
</div>
<div class="section level2">
<h2 id="set-up">Set-up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/filtered_sorted.bam"</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/filtered_sorted.bam.bai"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"timeout"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12751214/files/filtered_sorted.bam?download=1"</span>, <span class="st">"filtered_sorted.bam"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12751214/files/filtered_sorted.bam.bai?download=1"</span>, <span class="st">"filtered_sorted.bam.bai"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="va">bam</span> <span class="op">&lt;-</span> <span class="st">"filtered_sorted.bam"</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/path.expand.html" class="external-link">path.expand</a></span><span class="op">(</span><span class="st">"~/filtered_sorted.bam"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/subset_GRCh38.fa"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_47/GRCh38.primary_assembly.genome.fa.gz"</span>, <span class="st">"GRCh38.primary_assembly.genome.fa.gz"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="st">"GRCh38.primary_assembly.genome.fa.gz"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">" "</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">chr19</span> <span class="op">&lt;-</span> <span class="va">genome</span><span class="op">[</span><span class="va">genome</span><span class="op">$</span><span class="va">names</span> <span class="op">==</span> <span class="st">"chr19"</span><span class="op">]</span></span>
<span>  <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">writeXStringSet</a></span><span class="op">(</span><span class="va">chr19</span>, <span class="st">"~/subset_GRCh38.fa"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/path.expand.html" class="external-link">path.expand</a></span><span class="op">(</span><span class="st">"~/subset_GRCh38.fa"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/sce_lib10.qs"</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/sce_lib90.qs"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12770737/files/sce_lib10.qs?download=1"</span>, <span class="st">"sce_lib10.qs"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12770737/files/sce_lib90.qs?download=1"</span>, <span class="st">"sce_lib90.qs"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="va">sce_lib10</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"sce_lib10.qs"</span><span class="op">)</span></span>
<span>  <span class="va">sce_lib90</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"sce_lib90.qs"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">sce_lib10</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"~/sce_lib10.qs"</span><span class="op">)</span></span>
<span>  <span class="va">sce_lib90</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"~/sce_lib90.qs"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mritchielab.github.io/FLAMES" class="external-link">FLAMES</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="flames-pipeline">
<code>FLAMES</code> pipeline<a class="anchor" aria-label="anchor" href="#flames-pipeline"></a>
</h2>
<p><img src="FLAMES_pipeline.png" width="800"></p>
<div class="section level3">
<h3 id="read-pre-processing">Read pre-processing<a class="anchor" aria-label="anchor" href="#read-pre-processing"></a>
</h3>
<p>The <code>FLAMES</code> pipeline starts with de-multiplexing the
FASTQ reads, i.e. identifiying the cell barcode. For protocols that use
the 10x single-cell kits, the raw reads will look like the following,
where the 10x adapter sequences and polyT tails are highlighted. <img src="viewer_mode.png" width="800" alt="Screenshot from SeqSizzle"><br><img src="10x_3.png" width="800" alt="Chromium Single Cell 3â€™ Reagent Kits User Guide"></p>
<p><code>FLAMES</code> will locate the cell barcode by locating the
adjacent adapter sequences , and match the barcode against an allow-list
(if it is provided). We can specify the above 10x 3â€™ protocol as
follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">protocol_pattern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  primer <span class="op">=</span> <span class="st">"CTACACGACGCTCTTCCGATCT"</span>,</span>
<span>  BC <span class="op">=</span> <span class="st">"NNNNNNNNNNNNNNNN"</span>,</span>
<span>  UMI <span class="op">=</span> <span class="st">"NNNNNNNNNNNN"</span>, </span>
<span>  polyT <span class="op">=</span> <span class="st">"TTTTTTTTT"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that the order has to be the same as they appear in the
protocol. The names can be arbitrary for each sequence, but the names
<strong><code>UMI</code> and <code>BC</code> are required</strong> .</p>
<p>Letâ€™s try de-multiplexing a couple reads with
<code><a href="https://mritchielab.github.io/FLAMES/reference/find_barcode.html" class="external-link">FLAMES::find_barcode</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare a barcode allow-list</span></span>
<span><span class="va">bc_allow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_allow.tsv"</span><span class="op">)</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bc_allow.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>, </span>
<span>  destname <span class="op">=</span> <span class="va">bc_allow</span>, remove <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/find_barcode.html" class="external-link">find_barcode</a></span><span class="op">(</span></span>
<span>  <span class="co"># raw fastq file to be de-multiplexed</span></span>
<span>  fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  <span class="co"># outputs</span></span>
<span>  stats_out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_stat"</span><span class="op">)</span>,</span>
<span>  reads_out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>,</span>
<span>  barcodes_file <span class="op">=</span> <span class="va">bc_allow</span>,</span>
<span>  pattern <span class="op">=</span> <span class="va">protocol_pattern</span>, <span class="co"># our protocol pattern</span></span>
<span>  threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## FLEXIPLEX 0.96.2</span></span>
<span><span class="co">## Setting max barcode edit distance to 2</span></span>
<span><span class="co">## Setting max flanking sequence edit distance to 8</span></span>
<span><span class="co">## Setting read IDs to be  replaced</span></span>
<span><span class="co">## Setting number of threads to 1</span></span>
<span><span class="co">## Search pattern: </span></span>
<span><span class="co">## primer: CTACACGACGCTCTTCCGATCT</span></span>
<span><span class="co">## BC: NNNNNNNNNNNNNNNN</span></span>
<span><span class="co">## UMI: NNNNNNNNNNNN</span></span>
<span><span class="co">## polyT: TTTTTTTTT</span></span>
<span><span class="co">## Setting known barcodes from /tmp/RtmpQiGKkT/file96072f58bf4/bc_allow.tsv</span></span>
<span><span class="co">## Number of known barcodes: 143</span></span>
<span><span class="co">## Processing file: /usr/local/lib/R/site-library/FLAMES/extdata/fastq/demultiplexed.fq.gz</span></span>
<span><span class="co">## Searching for barcodes...</span></span>
<span><span class="co">## Processing file: /usr/local/lib/R/site-library/FLAMES/extdata/fastq/musc_rps24.fastq.gz</span></span>
<span><span class="co">## Searching for barcodes...</span></span>
<span><span class="co">## Number of reads processed: 403</span></span>
<span><span class="co">## Number of reads where at least one barcode was found: 368</span></span>
<span><span class="co">## Number of reads with exactly one barcode match: 364</span></span>
<span><span class="co">## Number of chimera reads: 1</span></span>
<span><span class="co">## All done!</span></span>
<span><span class="co">## Skipping TSO trimming...</span></span></code></pre>
<pre><code><span><span class="co">## $fastq</span></span>
<span><span class="co">## $fastq$reads_tb</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 372 Ã— 8</span></span></span>
<span><span class="co">##    Read            CellBarcode FlankEditDist BarcodeEditDist UMI   TooShort Outfile Sample</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 50409b8e-c5c8-â€¦ CGGGTCAGTAâ€¦             1               0 CCGGâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> d54ed65d-dad1-â€¦ CGGACTGAGTâ€¦             1               0 GGGAâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 94bfd5ea-64df-â€¦ CTGTGCTTCCâ€¦             1               2 TTCTâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> d6443706-7a0b-â€¦ GTCGTAATCCâ€¦             7               0 GGTAâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 055dd650-c268-â€¦ TTTATGCAGAâ€¦             1               0 ATTCâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 66cbd950-e2a9-â€¦ GTAACTGAGAâ€¦             2               0 TGAGâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> e427b9da-019d-â€¦ CGAACATGTCâ€¦             2               0 ATACâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 5f7420ae-9705-â€¦ ACCGTAACAAâ€¦             3               1 CCTCâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> f164d67e-d704-â€¦ GACTAACTCCâ€¦             3               1 GTTCâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> c7ae369e-ad72-â€¦ CATCGAACAGâ€¦             2               1 TTAAâ€¦ FALSE    /tmp/Râ€¦ fastq </span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 362 more rows</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $fastq$read_counts</span></span>
<span><span class="co">##                                       total reads </span></span>
<span><span class="co">##                                               403 </span></span>
<span><span class="co">##                               demultiplexed reads </span></span>
<span><span class="co">##                                               368 </span></span>
<span><span class="co">##                                single match reads </span></span>
<span><span class="co">##                                               364 </span></span>
<span><span class="co">## single strand single barcode multi-matching reads </span></span>
<span><span class="co">##                                                 0 </span></span>
<span><span class="co">##              single strand multiple barcode reads </span></span>
<span><span class="co">##                                                 3 </span></span>
<span><span class="co">##                 both strands single barcode reads </span></span>
<span><span class="co">##                                                 0 </span></span>
<span><span class="co">##               both strands multiple barcode reads </span></span>
<span><span class="co">##                                                 1</span></span></code></pre>
<p><code>FLAMES</code> will find and trim the specified pattern, save
the barcode and UMI to the identifier field of the read, let check the
first raw read:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>           n_max <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "@50409b8e-c5c8-4b86-8d2b-95effbb4434c"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   </span></span>
<span><span class="co">## [2] "GTTGCCGTTCAGTTGCTAAGCAGTGGTATCAACGCGATCCGCTGGGGCCTTGCGCGTTGATATGATTGGTCGGCAGTGAGCGGTCCTCTTTTCCTCCTCTCCCGGCTCGGCGCCGTAGCCATCATGAATGACAGTAACCATCCGGACCAGGAAGTTCATGACCAACCCAAATCTGCTCAGAGGAAACAGATGGTCATTGATATCCTTCATCCTGGGAAGGCAACAGTACCAAAAGGTGAAATTCAGGAAAAGCTGGCCAAAATGTACAAAACCACACCAGATGTCATCTTTGTATTTGGATTCAGAACCCACTTCAATTGGTAGCAAGACCACTAGCTTTGGCATGATCTATGATTCTTTAGATTATGCAAGAAGATGAGCCTAAACACAGACTGGCAAGACATGGCCTTTATGAGAGAAAGACCTCCGAAAACAGCAGAAAAGAGACACAAGAACAGAATGAAGAAGGTCAGGGGCACGGCGAGAGTAATGTTGGTGCTATGCAAAAAGTGAGTGGAGATTGGATACAGCCAAGGAGTAGATCTGCAGTGACTTGATGTTTGCTGTGATGTGCAGATTTCTGAGAGGACAAATAAACTAAAAGCTCCTACTAAAAAAAAAAAAAAAAAAAAAAGACAGCCCGGTACTACTACTGACCCGAGATCAGGAAGAGCGTCGTGTAGAGCAATACGTAAC"</span></span></code></pre>
<p>and the first de-multiplexed read:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>, n_max <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "@CGGGTCAGTAGTAGTA_CCGGGCTGTCTT#50409b8e-c5c8-4b86-8d2b-95effbb4434c_-1of1\tCB:Z:CGGGTCAGTAGTAGTA\tUB:Z:CCGGGCTGTCTT"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 </span></span>
<span><span class="co">## [2] "AGTAGGAGCTTTTAGTTTATTTGTCCTCTCAGAAATCTGCACATCACAGCAAACATCAAGTCACTGCAGATCTACTCCTTGGCTGTATCCAATCTCCACTCACTTTTTGCATAGCACCAACATTACTCTCGCCGTGCCCCTGACCTTCTTCATTCTGTTCTTGTGTCTCTTTTCTGCTGTTTTCGGAGGTCTTTCTCTCATAAAGGCCATGTCTTGCCAGTCTGTGTTTAGGCTCATCTTCTTGCATAATCTAAAGAATCATAGATCATGCCAAAGCTAGTGGTCTTGCTACCAATTGAAGTGGGTTCTGAATCCAAATACAAAGATGACATCTGGTGTGGTTTTGTACATTTTGGCCAGCTTTTCCTGAATTTCACCTTTTGGTACTGTTGCCTTCCCAGGATGAAGGATATCAATGACCATCTGTTTCCTCTGAGCAGATTTGGGTTGGTCATGAACTTCCTGGTCCGGATGGTTACTGTCATTCATGATGGCTACGGCGCCGAGCCGGGAGAGGAGGAAAAGAGGACCGCTCACTGCCGACCAATCATATCAACGCGCAAGGCCCCAGCGGATCGCGTTGATACCACTGCTTAGCAACTGAACGGCAAC"</span></span></code></pre>
<p>This was done with an Rcpp version of <em>flexiplex</em> <span class="citation">(<a href="#ref-chengFlexiplexVersatileDemultiplexer2024">Cheng <em>et
al.</em>, 2024-03-01</a>)</span>, and is equivalent to calling using the
native <code>flexiplex</code> command line tool with the following
arguments:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a> <span class="ex">~/flexiplex/flexiplex</span> <span class="at">-x</span> CTACACGACGCTCTTCCGATCT <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>  <span class="at">-b</span> <span class="st">'????????????????'</span> <span class="at">-u</span> <span class="st">'????????????'</span> <span class="at">-x</span> TTTTTTTTT <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  <span class="at">-e</span> 2 <span class="at">-f</span> 8 <span class="at">-k</span> bc_allow.tsv musc_rps24.fastq <span class="op">&gt;</span> main_out.fq</span></code></pre></div>
<p>If cell barcodes are not known in advance, e.g.Â when there is no
coupled short-reads, the <code>BLAZE</code> tool <span class="citation">(<a href="#ref-youIdentificationCellBarcodes2023">You
<em>et al.</em>, 2023-04-06</a>)</span> can be used to identify the cell
barcodes and UMI sequences , and <code>FLAMES</code> also provides a
wrapper <code><a href="https://mritchielab.github.io/FLAMES/reference/blaze.html" class="external-link">FLAMES::blaze</a></code> to call <code>BLAZE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/blaze.html" class="external-link">blaze</a></span><span class="op">(</span></span>
<span>  <span class="fl">100</span>,</span>
<span>  fq_in <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>Once the reads are processed, they are ready to be aligned to the
reference genome. <code>FLAMES</code> calls <code>minimap2</code> <span class="citation">(<a href="#ref-liMinimap2PairwiseAlignment2018">Li,
2018-09-15</a>)</span> to perform alignment and <code>samtoools</code>
<span class="citation">(<a href="#ref-liSequenceAlignmentMap2009">Li
<em>et al.</em>, 2009-08-15</a>)</span> to sort and index the BAM file,
so we need to make sure the binaries are available in <code>$PATH</code>
for <code>FLAMES</code> to find it.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/minimap2_align.html" class="external-link">minimap2_align</a></span><span class="op">(</span></span>
<span>    config <span class="op">=</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'config_sclr_nanopore_3end.json'</span>, package <span class="op">=</span> <span class="st">'FLAMES'</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    fa_file <span class="op">=</span> <span class="va">fa</span>,</span>
<span>    fq_in <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>,</span>
<span>    annot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"filtered.gtf"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span>,</span>
<span>    outdir <span class="op">=</span> <span class="va">outdir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>FLAMES</code> then performs gene quantification, isoform
identification and isoform quantification. For the interest of time we
will skip running them individually, as <code>FLAMES</code> also
provides the <code>sc_long_pipeline</code> wrapper to do all steps. It
will return a <code>SingleCellExperiment</code> object with the isoform
counts, and the isoform annotations stored in the <code>rowRanges</code>
slot. Letâ€™s run the pipeline with a small number of reads:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/sc_long_pipeline.html" class="external-link">sc_long_pipeline</a></span><span class="op">(</span></span>
<span>  genome_fa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.fa.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.gtf.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>  barcodes_file <span class="op">=</span> <span class="va">bc_allow</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h2>
<p>After running the pipeline, we get a
<code>SingleCellExperiment</code> object and we can finally start some
analysis. Letâ€™s load the <code>SingleCellExperiment</code> object from
our previous cell line mixture data <span class="citation">(<a href="#ref-tianComprehensiveCharacterizationSinglecell2021">Tian <em>et
al.</em>, 2021-11-11</a>)</span>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce.qs"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 15076 162 </span></span>
<span><span class="co">## metadata(1): OutputFiles</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(15076): ENSG00000000419.14_50934870_50958532_1</span></span>
<span><span class="co">##   ENSG00000001036.14_143495487_143511720_2 ... ENST00000713574.1</span></span>
<span><span class="co">##   ENST00000713579.1</span></span>
<span><span class="co">## rowData names(3): transcript_id gene_id FSM_match</span></span>
<span><span class="co">## colnames(162): AAAGAACAGCGATCGA AACAGGGTCCAGTGCG ... TTTCAGTCACCAGTAT</span></span>
<span><span class="co">##   TTTGGAGCACAAAGTA</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(1): gene</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 14555 162 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(14555): ENSG00000179820.16 ENSG00000114062.22 ... ENSG00000212907.2</span></span>
<span><span class="co">##   ENSG00000228253.1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(162): AAAGAACAGCGATCGA AACAGGGTCCAGTGCG ... TTTCAGTCACCAGTAT</span></span>
<span><span class="co">##   TTTGGAGCACAAAGTA</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>I have also attached the gene count matrix from long-reads to the
<code>altExp</code> slot. Letâ€™s do clustering with the gene counts,
using Bioconductor packages for normalization, dimensionality reduction
and clustering:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">cluster_leiden</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/communities.html" class="external-link">membership</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'sizeFactor'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing-2.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sce &lt;- qs::qread(system.file("extdata", "processed_sce.qs", package = "IgniteRNAseq"))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/ChangqingW/IgniteRNAseq/devel/inst/extdata/processed_sce.qs"</span>, <span class="st">"processed_sce.qs"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="st">"processed_sce.qs"</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'sizeFactor'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing2-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing2-2.png" width="700"></p>
<p>We can clearly see the five cell lines are identified by the
clustering using the gene counts from the long-reads. Now letâ€™s see if
we can get mutations / isoforms specific to each cell line.</p>
</div>
<div class="section level2">
<h2 id="mutation-analysis">Mutation analysis<a class="anchor" aria-label="anchor" href="#mutation-analysis"></a>
</h2>
<div class="section level3">
<h3 id="mutation-discovery">Mutation discovery<a class="anchor" aria-label="anchor" href="#mutation-discovery"></a>
</h3>
<p>Letâ€™s suppose we no nothing about the mutations in our cell lines, we
can use the <code><a href="https://mritchielab.github.io/FLAMES/reference/find_variants.html" class="external-link">FLAMES::find_variants</a></code> function get bulk
mutation count (similar to the <code>samtools mpileup</code> command,
but in a tibble format with extra information):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/find_variants.html" class="external-link">find_variants</a></span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">fa</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"filtered.gtf"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span>,</span>
<span>  min_nucleotide_depth <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  annotated_region_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names_from <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 07:52:45 Reading reference ...</span></span></code></pre>
<pre><code><span><span class="co">## 07:52:45 Reading annotation ...</span></span></code></pre>
<pre><code><span><span class="co">## 07:52:45 Got 1 bam file, parallelizing over each region ...</span></span></code></pre>
<pre><code><span><span class="co">##   |                                                                                        |                                                                                |   0%  |                                                                                        |================                                                                |  20%  |                                                                                        |================================                                                |  40%  |                                                                                        |================================================                                |  60%  |                                                                                        |================================================================                |  80%  |                                                                                        |================================================================================| 100%</span></span></code></pre>
<pre><code><span><span class="co">## 07:53:35 Merging results ...</span></span></code></pre>
<pre><code><span><span class="co">## 07:53:35 Calculating homopolymer percentages ...</span></span></code></pre>
<pre><code><span><span class="co">##   |                                                                                        |                                                                                |   0%  |                                                                                        |                                                                                |   1%  |                                                                                        |=                                                                               |   1%  |                                                                                        |=                                                                               |   2%  |                                                                                        |==                                                                              |   2%  |                                                                                        |==                                                                              |   3%  |                                                                                        |===                                                                             |   3%  |                                                                                        |===                                                                             |   4%  |                                                                                        |====                                                                            |   4%  |                                                                                        |====                                                                            |   5%  |                                                                                        |====                                                                            |   6%  |                                                                                        |=====                                                                           |   6%  |                                                                                        |=====                                                                           |   7%  |                                                                                        |======                                                                          |   7%  |                                                                                        |======                                                                          |   8%  |                                                                                        |=======                                                                         |   8%  |                                                                                        |=======                                                                         |   9%  |                                                                                        |========                                                                        |   9%  |                                                                                        |========                                                                        |  10%  |                                                                                        |========                                                                        |  11%  |                                                                                        |=========                                                                       |  11%  |                                                                                        |=========                                                                       |  12%  |                                                                                        |==========                                                                      |  12%  |                                                                                        |==========                                                                      |  13%  |                                                                                        |===========                                                                     |  13%  |                                                                                        |===========                                                                     |  14%  |                                                                                        |============                                                                    |  15%  |                                                                                        |============                                                                    |  16%  |                                                                                        |=============                                                                   |  16%  |                                                                                        |=============                                                                   |  17%  |                                                                                        |==============                                                                  |  17%  |                                                                                        |==============                                                                  |  18%  |                                                                                        |===============                                                                 |  18%  |                                                                                        |===============                                                                 |  19%  |                                                                                        |================                                                                |  20%  |                                                                                        |================                                                                |  21%  |                                                                                        |=================                                                               |  21%  |                                                                                        |=================                                                               |  22%  |                                                                                        |==================                                                              |  22%  |                                                                                        |==================                                                              |  23%  |                                                                                        |===================                                                             |  23%  |                                                                                        |===================                                                             |  24%  |                                                                                        |====================                                                            |  25%  |                                                                                        |====================                                                            |  26%  |                                                                                        |=====================                                                           |  26%  |                                                                                        |=====================                                                           |  27%  |                                                                                        |======================                                                          |  27%  |                                                                                        |======================                                                          |  28%  |                                                                                        |=======================                                                         |  28%  |                                                                                        |=======================                                                         |  29%  |                                                                                        |========================                                                        |  30%  |                                                                                        |========================                                                        |  31%  |                                                                                        |=========================                                                       |  31%  |                                                                                        |=========================                                                       |  32%  |                                                                                        |==========================                                                      |  32%  |                                                                                        |==========================                                                      |  33%  |                                                                                        |===========================                                                     |  33%  |                                                                                        |===========================                                                     |  34%  |                                                                                        |============================                                                    |  35%  |                                                                                        |============================                                                    |  36%  |                                                                                        |=============================                                                   |  36%  |                                                                                        |=============================                                                   |  37%  |                                                                                        |==============================                                                  |  37%  |                                                                                        |==============================                                                  |  38%  |                                                                                        |===============================                                                 |  38%  |                                                                                        |===============================                                                 |  39%  |                                                                                        |================================                                                |  40%  |                                                                                        |================================                                                |  41%  |                                                                                        |=================================                                               |  41%  |                                                                                        |=================================                                               |  42%  |                                                                                        |==================================                                              |  42%  |                                                                                        |==================================                                              |  43%  |                                                                                        |===================================                                             |  43%  |                                                                                        |===================================                                             |  44%  |                                                                                        |====================================                                            |  45%  |                                                                                        |====================================                                            |  46%  |                                                                                        |=====================================                                           |  46%  |                                                                                        |=====================================                                           |  47%  |                                                                                        |======================================                                          |  47%  |                                                                                        |======================================                                          |  48%  |                                                                                        |=======================================                                         |  48%  |                                                                                        |=======================================                                         |  49%  |                                                                                        |========================================                                        |  50%  |                                                                                        |=========================================                                       |  51%  |                                                                                        |=========================================                                       |  52%  |                                                                                        |==========================================                                      |  52%  |                                                                                        |==========================================                                      |  53%  |                                                                                        |===========================================                                     |  53%  |                                                                                        |===========================================                                     |  54%  |                                                                                        |============================================                                    |  54%  |                                                                                        |============================================                                    |  55%  |                                                                                        |=============================================                                   |  56%  |                                                                                        |=============================================                                   |  57%  |                                                                                        |==============================================                                  |  57%  |                                                                                        |==============================================                                  |  58%  |                                                                                        |===============================================                                 |  58%  |                                                                                        |===============================================                                 |  59%  |                                                                                        |================================================                                |  59%  |                                                                                        |================================================                                |  60%  |                                                                                        |=================================================                               |  61%  |                                                                                        |=================================================                               |  62%  |                                                                                        |==================================================                              |  62%  |                                                                                        |==================================================                              |  63%  |                                                                                        |===================================================                             |  63%  |                                                                                        |===================================================                             |  64%  |                                                                                        |====================================================                            |  64%  |                                                                                        |====================================================                            |  65%  |                                                                                        |=====================================================                           |  66%  |                                                                                        |=====================================================                           |  67%  |                                                                                        |======================================================                          |  67%  |                                                                                        |======================================================                          |  68%  |                                                                                        |=======================================================                         |  68%  |                                                                                        |=======================================================                         |  69%  |                                                                                        |========================================================                        |  69%  |                                                                                        |========================================================                        |  70%  |                                                                                        |=========================================================                       |  71%  |                                                                                        |=========================================================                       |  72%  |                                                                                        |==========================================================                      |  72%  |                                                                                        |==========================================================                      |  73%  |                                                                                        |===========================================================                     |  73%  |                                                                                        |===========================================================                     |  74%  |                                                                                        |============================================================                    |  74%  |                                                                                        |============================================================                    |  75%  |                                                                                        |=============================================================                   |  76%  |                                                                                        |=============================================================                   |  77%  |                                                                                        |==============================================================                  |  77%  |                                                                                        |==============================================================                  |  78%  |                                                                                        |===============================================================                 |  78%  |                                                                                        |===============================================================                 |  79%  |                                                                                        |================================================================                |  79%  |                                                                                        |================================================================                |  80%  |                                                                                        |=================================================================               |  81%  |                                                                                        |=================================================================               |  82%  |                                                                                        |==================================================================              |  82%  |                                                                                        |==================================================================              |  83%  |                                                                                        |===================================================================             |  83%  |                                                                                        |===================================================================             |  84%  |                                                                                        |====================================================================            |  84%  |                                                                                        |====================================================================            |  85%  |                                                                                        |=====================================================================           |  86%  |                                                                                        |=====================================================================           |  87%  |                                                                                        |======================================================================          |  87%  |                                                                                        |======================================================================          |  88%  |                                                                                        |=======================================================================         |  88%  |                                                                                        |=======================================================================         |  89%  |                                                                                        |========================================================================        |  89%  |                                                                                        |========================================================================        |  90%  |                                                                                        |========================================================================        |  91%  |                                                                                        |=========================================================================       |  91%  |                                                                                        |=========================================================================       |  92%  |                                                                                        |==========================================================================      |  92%  |                                                                                        |==========================================================================      |  93%  |                                                                                        |===========================================================================     |  93%  |                                                                                        |===========================================================================     |  94%  |                                                                                        |============================================================================    |  94%  |                                                                                        |============================================================================    |  95%  |                                                                                        |============================================================================    |  96%  |                                                                                        |=============================================================================   |  96%  |                                                                                        |=============================================================================   |  97%  |                                                                                        |==============================================================================  |  97%  |                                                                                        |==============================================================================  |  98%  |                                                                                        |=============================================================================== |  98%  |                                                                                        |=============================================================================== |  99%  |                                                                                        |================================================================================|  99%  |                                                                                        |================================================================================| 100%</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 501 Ã— 10</span></span></span>
<span><span class="co">##    seqnames   pos nucleotide  count    sum    freq ref     bam_path region homopolymer_pct</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> chrM      <span style="text-decoration: underline;">7</span>020 G          <span style="text-decoration: underline;">245</span>390 <span style="text-decoration: underline;">245</span>390 1       characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> chrM      <span style="text-decoration: underline;">7</span>021 -            <span style="text-decoration: underline;">2</span>293 <span style="text-decoration: underline;">247</span>907 0.009<span style="text-decoration: underline;">25</span> characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> chrM      <span style="text-decoration: underline;">7</span>021 C            <span style="text-decoration: underline;">2</span>509 <span style="text-decoration: underline;">247</span>907 0.010<span style="text-decoration: underline;">1</span>  characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> chrM      <span style="text-decoration: underline;">7</span>021 T          <span style="text-decoration: underline;">243</span>105 <span style="text-decoration: underline;">247</span>907 0.981   characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> chrM      <span style="text-decoration: underline;">7</span>023 +            <span style="text-decoration: underline;">2</span>244 <span style="text-decoration: underline;">247</span>146 0.009<span style="text-decoration: underline;">08</span> characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> chrM      <span style="text-decoration: underline;">7</span>022 -            <span style="text-decoration: underline;">2</span>375 <span style="text-decoration: underline;">247</span>930 0.009<span style="text-decoration: underline;">58</span> characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> chrM      <span style="text-decoration: underline;">7</span>022 T          <span style="text-decoration: underline;">245</span>555 <span style="text-decoration: underline;">247</span>930 0.990   characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> chrM      <span style="text-decoration: underline;">7</span>024 +            <span style="text-decoration: underline;">3</span>434 <span style="text-decoration: underline;">248</span>057 0.013<span style="text-decoration: underline;">8</span>  characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> chrM      <span style="text-decoration: underline;">7</span>023 -            <span style="text-decoration: underline;">2</span>929 <span style="text-decoration: underline;">247</span>146 0.011<span style="text-decoration: underline;">9</span>  characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> chrM      <span style="text-decoration: underline;">7</span>023 G          <span style="text-decoration: underline;">244</span>217 <span style="text-decoration: underline;">247</span>146 0.988   characâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 491 more rows</span></span></span></code></pre>
<p>Since this is Nanopore sequencing, there ones with very low
alternative allele frequency are likely to be sequencing errors;
Nanopore is also known to have higher error rate in homopolymer regions,
so we can filter out those as well:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span> <span class="op">&lt;-</span> <span class="va">mutation_discovery_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">freq</span> <span class="op">&gt;</span> <span class="fl">0.2</span>, <span class="va">freq</span> <span class="op">&lt;</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mutation_discovery_tb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 12 Ã— 10</span></span></span>
<span><span class="co">##    seqnames      pos nucleotide  count    sum  freq ref    bam_path region homopolymer_pct</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> chrM         <span style="text-decoration: underline;">8</span>228 C          <span style="text-decoration: underline;">211</span>448 <span style="text-decoration: underline;">266</span>121 0.795 charaâ€¦ filtereâ€¦ MT-CO2            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> chrM         <span style="text-decoration: underline;">8</span>269 G          <span style="text-decoration: underline;">209</span>767 <span style="text-decoration: underline;">264</span>311 0.794 charaâ€¦ filtereâ€¦ MT-CO2            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> chrM         <span style="text-decoration: underline;">8</span>251 G          <span style="text-decoration: underline;">156</span>306 <span style="text-decoration: underline;">263</span>505 0.593 charaâ€¦ filtereâ€¦ MT-CO2            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> chrM         <span style="text-decoration: underline;">9</span>123 G          <span style="text-decoration: underline;">150</span>286 <span style="text-decoration: underline;">219</span>009 0.686 charaâ€¦ filtereâ€¦ MT-ATâ€¦            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> chrM         <span style="text-decoration: underline;">7</span>028 T          <span style="text-decoration: underline;">130</span>634 <span style="text-decoration: underline;">249</span>199 0.524 charaâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> chrM         <span style="text-decoration: underline;">7</span>028 C          <span style="text-decoration: underline;">115</span>940 <span style="text-decoration: underline;">249</span>199 0.465 charaâ€¦ filtereâ€¦ MT-CO1            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> chrM         <span style="text-decoration: underline;">8</span>251 A          <span style="text-decoration: underline;">104</span>893 <span style="text-decoration: underline;">263</span>505 0.398 charaâ€¦ filtereâ€¦ MT-CO2            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> chrM        <span style="text-decoration: underline;">11</span>719 G           <span style="text-decoration: underline;">98</span>511 <span style="text-decoration: underline;">149</span>795 0.658 charaâ€¦ filtereâ€¦ MT-ND4            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> chr19    48<span style="text-decoration: underline;">965</span>830 C           <span style="text-decoration: underline;">85</span>423 <span style="text-decoration: underline;">178</span>263 0.479 charaâ€¦ filtereâ€¦ FTL               -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> chr19    48<span style="text-decoration: underline;">965</span>830 T           <span style="text-decoration: underline;">84</span>346 <span style="text-decoration: underline;">178</span>263 0.473 charaâ€¦ filtereâ€¦ FTL               -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> chrM         <span style="text-decoration: underline;">9</span>123 A           <span style="text-decoration: underline;">66</span>376 <span style="text-decoration: underline;">219</span>009 0.303 charaâ€¦ filtereâ€¦ MT-ATâ€¦            -<span style="color: #BB0000;">Inf</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> chrM        <span style="text-decoration: underline;">11</span>719 A           <span style="text-decoration: underline;">51</span>284 <span style="text-decoration: underline;">149</span>795 0.342 charaâ€¦ filtereâ€¦ MT-ND4            -<span style="color: #BB0000;">Inf</span></span></span></code></pre>
<p>As we can see, there is a couple of potential SNP mutations, letâ€™s
see if they are correlated with our clustering.</p>
</div>
<div class="section level3">
<h3 id="mutation-calling">Mutation calling<a class="anchor" aria-label="anchor" href="#mutation-calling"></a>
</h3>
<p>The previous <code>find_variants</code> function is for bulk mutation
calling, but since we have <strong>single-cell long-reads</strong>, we
can call mutations at the single-cell level, and associate them with
transcription profiles.</p>
<p><code>FLAMES</code> offers the <code>sc_mutations</code> function to
call mutations at the single-cell level:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_tb</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/sc_mutations.html" class="external-link">sc_mutations</a></span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">mutation_discovery_tb</span><span class="op">$</span><span class="va">seqnames</span><span class="op">)</span>,</span>
<span>  positions <span class="op">=</span> <span class="va">mutation_discovery_tb</span><span class="op">$</span><span class="va">pos</span>,</span>
<span>  indel <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  barcodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 07:54:00 Got 1 bam file, parallelizing over each position ...</span></span></code></pre>
<pre><code><span><span class="co">##   |                                                                                        |                                                                                |   0%  |                                                                                        |=======                                                                         |   8%  |                                                                                        |=============                                                                   |  17%  |                                                                                        |====================                                                            |  25%  |                                                                                        |===========================                                                     |  33%  |                                                                                        |=================================                                               |  42%  |                                                                                        |========================================                                        |  50%  |                                                                                        |===============================================                                 |  58%  |                                                                                        |=====================================================                           |  67%  |                                                                                        |============================================================                    |  75%  |                                                                                        |===================================================================             |  83%  |                                                                                        |=========================================================================       |  92%  |                                                                                        |================================================================================| 100%</span></span></code></pre>
<pre><code><span><span class="co">## 07:55:38 Merging results ...</span></span></code></pre>
<p>Lets visualize the SNP status of each cell on the long-read derived
UMAP. We will color the cell by the allele that is supported by more
than half of the reads from that cell. This can be done easily with some
tidyverse functions. The cell lines are homozygous, but with e.g.Â cancer
data this could require more considerations.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snps_tb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 Ã— 7</span></span></span>
<span><span class="co">##   allele barcode          allele_count cell_total_reads     pct   pos seqname</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> A      AAAGAACAGCGATCGA            1              828 0.001<span style="text-decoration: underline;">21</span>  <span style="text-decoration: underline;">8</span>228 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> A      AACAGGGTCCAGTGCG            0             <span style="text-decoration: underline;">1</span>109 0        <span style="text-decoration: underline;">8</span>228 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> A      AAGTCGTCAGTAGGAC            3             <span style="text-decoration: underline;">2</span>483 0.001<span style="text-decoration: underline;">21</span>  <span style="text-decoration: underline;">8</span>228 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> A      AATCACGGTCCTCCTA            0             <span style="text-decoration: underline;">1</span>538 0        <span style="text-decoration: underline;">8</span>228 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> A      AATCGACAGGGTGGGA            0             <span style="text-decoration: underline;">1</span>325 0        <span style="text-decoration: underline;">8</span>228 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> A      AATCGACTCGTTTACT            0             <span style="text-decoration: underline;">1</span>368 0        <span style="text-decoration: underline;">8</span>228 chrM</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr19</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chr19'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">48965830</span><span class="op">)</span></span>
<span><span class="va">chr19</span> <span class="op">&lt;-</span> <span class="va">chr19</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chr19</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chr19</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-1.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_7028</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">7028</span><span class="op">)</span></span>
<span><span class="va">chrM_7028</span> <span class="op">&lt;-</span> <span class="va">chrM_7028</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_7028</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_7028</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-2.png" width="700"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_8251</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">8251</span><span class="op">)</span></span>
<span><span class="va">chrM_8251</span> <span class="op">&lt;-</span> <span class="va">chrM_8251</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_8251</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_8251</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-3.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_9123</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">9123</span><span class="op">)</span></span>
<span><span class="va">chrM_9123</span> <span class="op">&lt;-</span> <span class="va">chrM_9123</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_9123</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_9123</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-4.png" width="700"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_11719</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">11719</span><span class="op">)</span></span>
<span><span class="va">chrM_11719</span> <span class="op">&lt;-</span> <span class="va">chrM_11719</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_11719</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_11719</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-5.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-transcript-experssion-analysis">Differential Transcript Experssion analysis<a class="anchor" aria-label="anchor" href="#differential-transcript-experssion-analysis"></a>
</h2>
<p>Since we have the transcript count matrix from <code>FLAMES</code>,
we can do some DTE analysis. Letâ€™s pretend that instead of cell lines,
we have two replicate samples, and we devide the cells in each sample by
a SNP mutation status. We can use <code>limma</code> to find
differentially transcribed genes in the mutated vs reference cells.</p>
<p>First we need to aggregate the counts to pseudo-bulk samples:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># aggregate to pseudo-bulk counts</span></span>
<span><span class="va">pseudo_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">colSums</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">identity</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="va">colSums</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">identity</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span></span>
<span>  <span class="co"># throw out one cell line to make this 2 vs 2</span></span>
<span>  counts <span class="op">=</span> <span class="va">pseudo_bulk</span><span class="op">[</span>, <span class="op">-</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mutated"</span>, <span class="st">"reference"</span>, <span class="st">"mutated"</span>, <span class="st">"reference"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove lowly expressed isoforms</span></span>
<span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="va">dgelist</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dgelist</span><span class="op">)</span>, , </span>
<span>  keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize the pseudo-bulk samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="va">dgelist</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/MDS-1.png" width="700"></p>
<p>Looks like there is some within-group variation as well as
between-group variation. Letâ€™s use <code>limma</code> to find
significantly differentiated expressed transcripts between the mutated
and reference cells.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dgelist</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span>
<span><span class="co"># means model</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dgelist</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span>, data <span class="op">=</span> <span class="va">dgelist</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*\\$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" .*$"</span>, <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span></span>
<span><span class="va">dge_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dgelist</span>, <span class="va">design</span>, save.plot <span class="op">=</span> <span class="cn">T</span>, plot <span class="op">=</span> <span class="cn">F</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">dge_v</span>, <span class="va">dge_v</span><span class="op">$</span><span class="va">design</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span></span>
<span>    MalevsFemale <span class="op">=</span> <span class="st">"groupmutated - groupreference"</span>,</span>
<span>    levels <span class="op">=</span> <span class="va">dge_v</span><span class="op">$</span><span class="va">design</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    logFC AveExpr      t  P.Value adj.P.Val    B</span></span>
<span><span class="co">## ENST00000621160.5  -7.20    7.01 -11.71 1.29e-06    0.0103 5.15</span></span>
<span><span class="co">## ENST00000284440.9   5.95    6.64  11.00 2.15e-06    0.0103 4.92</span></span>
<span><span class="co">## ENST00000618200.4  -7.91    6.81 -11.03 2.10e-06    0.0103 4.46</span></span>
<span><span class="co">## ENST00000649529.1  -5.55    7.65  -9.11 9.85e-06    0.0289 3.80</span></span>
<span><span class="co">## ENST00000316403.15  5.12    5.99   8.87 1.21e-05    0.0289 3.58</span></span>
<span><span class="co">## ENST00000412585.7  -4.57    6.00  -8.60 1.55e-05    0.0306 3.37</span></span>
<span><span class="co">## ENST00000361157.11 -4.61    6.96  -8.41 1.85e-05    0.0306 3.27</span></span>
<span><span class="co">## ENST00000361566.7  -5.38    8.03  -8.29 2.06e-05    0.0306 3.19</span></span>
<span><span class="co">## ENST00000378292.9   6.48    5.22   9.03 1.05e-05    0.0289 3.18</span></span>
<span><span class="co">## ENST00000368716.9   4.90    6.44   8.25 2.14e-05    0.0306 3.14</span></span></code></pre>
<p>Letâ€™s visualise the top differentially expressed transcript on the
UMAP:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">]</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>exp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">'ENST00000621160.5'</span>, <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  shape_by <span class="op">=</span>  <span class="st">"cluster"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="differential-transcript-usage-analysis">Differential Transcript Usage analysis<a class="anchor" aria-label="anchor" href="#differential-transcript-usage-analysis"></a>
</h2>
<p>Previously we have identified the transcripts that are differentially
expressed ( i.e.Â DTE), another question we can ask is whether there are
differentiall transcript usage (DTU) between the mutated and reference
cells. The difference between DTU and DTE is illustrated by the
following figure, adapted from <span class="citation">(<a href="#ref-ntranosDiscriminativeLearningApproach2019">Ntranos <em>et
al.</em>, 2019-02</a>)</span>. <img src="DTU_DTE.png" width="800"></p>
<p>The <code>sc_DTU_analysis</code> function can test for DTU (chi-squre
test against uniform distribution). It will ignore the transcription
start site and end site differences and only consider splicing
differences.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The splice sites are saved in isoform_FSM_annotation.csv during</span></span>
<span><span class="co"># isoform identification. Let's make sure it is available:</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">OutputFiles</span><span class="op">$</span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://zenodo.org/records/12751214/files/isoform_FSM_annotation.csv?download=1'</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">sce</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">OutputFiles</span><span class="op">$</span><span class="va">outdir</span>, <span class="st">"isoform_FSM_annotation.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">dtu_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/sc_DTU_analysis.html" class="external-link">sc_DTU_analysis</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p>We will get a data frame with the chi-square test results:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dtu_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 gene_id X_value df                                   DTU_tr DTU_group</span></span>
<span><span class="co">## 794  ENSG00000130066.17    2311 12   ENSG00000130066.17_23783173_23799848_2         4</span></span>
<span><span class="co">## 1634 ENSG00000186468.13    1733 12   ENSG00000186468.13_82276060_82278355_2         1</span></span>
<span><span class="co">## 2229 ENSG00000170889.14    1691 12                        ENST00000391751.7         4</span></span>
<span><span class="co">## 2477 ENSG00000138326.21    6555 12                        ENST00000360830.9         5</span></span>
<span><span class="co">## 2623 ENSG00000243678.12    4943 12                        ENST00000503064.5         2</span></span>
<span><span class="co">## 1127 ENSG00000148346.12    1027 12 ENSG00000148346.12_128149453_128153453_7         5</span></span>
<span><span class="co">##        p_value     adj_p</span></span>
<span><span class="co">## 794   0.00e+00  0.00e+00</span></span>
<span><span class="co">## 1634  0.00e+00  0.00e+00</span></span>
<span><span class="co">## 2229  0.00e+00  0.00e+00</span></span>
<span><span class="co">## 2477  0.00e+00  0.00e+00</span></span>
<span><span class="co">## 2623  0.00e+00  0.00e+00</span></span>
<span><span class="co">## 1127 2.81e-212 7.95e-209</span></span></code></pre>
<div class="section level3">
<h3 id="visualizations-for-dtu">Visualizations for DTU<a class="anchor" aria-label="anchor" href="#visualizations-for-dtu"></a>
</h3>
<p>Unlike DTE, DTU is a bit harder to visiuliaze. For experiments
similar to <code>FLT-seq</code>, where matching short-reads are
available for both the smaller sub-sample and the larger sub-sample, we
provide some additional plotting functions:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiAssay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mritchielab.github.io/FLAMES/reference/combine_sce.html" class="external-link">combine_sce</a></span><span class="op">(</span></span>
<span>  <span class="va">sce_lib90</span>,</span>
<span>  <span class="va">sce_lib10</span>,</span>
<span>  <span class="va">sce</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">transcript_id</span> <span class="op">==</span><span class="st">'ENST00000391751.7'</span>, <span class="op">]</span></span>
<span><span class="fu">sc_umap_expression</span><span class="op">(</span><span class="st">"ENSG00000170889.14"</span>, <span class="va">multiAssay</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] edgeR_4.4.0                 limma_3.62.0                lubridate_1.9.3            </span></span>
<span><span class="co">##  [4] forcats_1.0.0               stringr_1.5.1               dplyr_1.1.4                </span></span>
<span><span class="co">##  [7] purrr_1.0.2                 readr_2.1.5                 tidyr_1.3.1                </span></span>
<span><span class="co">## [10] tibble_3.2.1                tidyverse_2.0.0             scater_1.34.0              </span></span>
<span><span class="co">## [13] ggplot2_3.5.1               scuttle_1.16.0              SingleCellExperiment_1.28.0</span></span>
<span><span class="co">## [16] SummarizedExperiment_1.36.0 Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">## [19] GenomeInfoDb_1.42.0         IRanges_2.40.0              S4Vectors_0.44.0           </span></span>
<span><span class="co">## [22] BiocGenerics_0.52.0         MatrixGenerics_1.18.0       matrixStats_1.4.1          </span></span>
<span><span class="co">## [25] FLAMES_2.0.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.5                  ProtGenerics_1.38.0       bitops_1.0-9             </span></span>
<span><span class="co">##   [4] httr_1.4.7                RColorBrewer_1.1-3        doParallel_1.0.17        </span></span>
<span><span class="co">##   [7] tools_4.4.1               backports_1.5.0           utf8_1.2.4               </span></span>
<span><span class="co">##  [10] R6_2.5.1                  HDF5Array_1.34.0          uwot_0.2.2               </span></span>
<span><span class="co">##  [13] lazyeval_0.2.2            rhdf5filters_1.18.0       GetoptLong_1.0.5         </span></span>
<span><span class="co">##  [16] withr_3.0.2               prettyunits_1.2.0         GGally_2.2.1             </span></span>
<span><span class="co">##  [19] gridExtra_2.3             cli_3.6.3                 textshaping_0.4.0        </span></span>
<span><span class="co">##  [22] scatterpie_0.2.4          labeling_0.4.3            ggbio_1.54.0             </span></span>
<span><span class="co">##  [25] sass_0.4.9                pkgdown_2.1.1             yulab.utils_0.1.7        </span></span>
<span><span class="co">##  [28] Rsamtools_2.22.0          systemfonts_1.1.0         txdbmaker_1.2.0          </span></span>
<span><span class="co">##  [31] foreign_0.8-87            R.utils_2.12.3            dichromat_2.0-0.1        </span></span>
<span><span class="co">##  [34] parallelly_1.38.0         BSgenome_1.74.0           rstudioapi_0.17.1        </span></span>
<span><span class="co">##  [37] RSQLite_2.3.7             FNN_1.1.4.1               generics_0.1.3           </span></span>
<span><span class="co">##  [40] shape_1.4.6.1             RApiSerialize_0.1.4       BiocIO_1.16.0            </span></span>
<span><span class="co">##  [43] vroom_1.6.5               Matrix_1.7-1              ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##  [46] fansi_1.0.6               abind_1.4-8               R.methodsS3_1.8.2        </span></span>
<span><span class="co">##  [49] lifecycle_1.0.4           yaml_2.3.10               rhdf5_2.50.0             </span></span>
<span><span class="co">##  [52] SparseArray_1.6.0         BiocFileCache_2.14.0      grid_4.4.1               </span></span>
<span><span class="co">##  [55] blob_1.2.4                dqrng_0.4.1               crayon_1.5.3             </span></span>
<span><span class="co">##  [58] dir.expiry_1.14.0         lattice_0.22-6            beachmat_2.22.0          </span></span>
<span><span class="co">##  [61] cowplot_1.1.3             GenomicFeatures_1.58.0    KEGGREST_1.46.0          </span></span>
<span><span class="co">##  [64] magick_2.8.5              metapod_1.14.0            pillar_1.9.0             </span></span>
<span><span class="co">##  [67] knitr_1.48                ComplexHeatmap_2.22.0     bambu_3.8.0              </span></span>
<span><span class="co">##  [70] rjson_0.2.23              xgboost_1.7.8.1           codetools_0.2-20         </span></span>
<span><span class="co">##  [73] glue_1.8.0                ggfun_0.1.7               data.table_1.16.2        </span></span>
<span><span class="co">##  [76] vctrs_0.6.5               png_0.1-8                 gtable_0.3.6             </span></span>
<span><span class="co">##  [79] cachem_1.1.0              xfun_0.48                 S4Arrays_1.6.0           </span></span>
<span><span class="co">##  [82] DropletUtils_1.26.0       iterators_1.0.14          bluster_1.16.0           </span></span>
<span><span class="co">##  [85] statmod_1.5.0             bit64_4.5.2               progress_1.2.3           </span></span>
<span><span class="co">##  [88] filelock_1.0.3            bslib_0.8.0               irlba_2.3.5.1            </span></span>
<span><span class="co">##  [91] vipor_0.4.7               rpart_4.1.23              colorspace_2.1-1         </span></span>
<span><span class="co">##  [94] DBI_1.2.3                 Hmisc_5.2-0               nnet_7.3-19              </span></span>
<span><span class="co">##  [97] tidyselect_1.2.1          bit_4.5.0                 compiler_4.4.1           </span></span>
<span><span class="co">## [100] curl_5.2.3                httr2_1.0.5               graph_1.84.0             </span></span>
<span><span class="co">## [103] htmlTable_2.4.3           BiocNeighbors_2.0.0       basilisk.utils_1.18.0    </span></span>
<span><span class="co">## [106] xml2_1.3.6                desc_1.4.3                DelayedArray_0.32.0      </span></span>
<span><span class="co">## [109] stringfish_0.16.0         rtracklayer_1.66.0        checkmate_2.3.2          </span></span>
<span><span class="co">## [112] scales_1.3.0              RBGL_1.82.0               rappdirs_0.3.3           </span></span>
<span><span class="co">## [115] SpatialExperiment_1.16.0  digest_0.6.37             rmarkdown_2.28           </span></span>
<span><span class="co">## [118] basilisk_1.18.0           XVector_0.46.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">## [121] pkgconfig_2.0.3           base64enc_0.1-3           sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">## [124] highr_0.11                dbplyr_2.5.0              fastmap_1.2.0            </span></span>
<span><span class="co">## [127] ensembldb_2.30.0          rlang_1.1.4               GlobalOptions_0.1.2      </span></span>
<span><span class="co">## [130] htmlwidgets_1.6.4         UCSC.utils_1.2.0          DelayedMatrixStats_1.28.0</span></span>
<span><span class="co">## [133] farver_2.1.2              jquerylib_0.1.4           jsonlite_1.8.9           </span></span>
<span><span class="co">## [136] BiocParallel_1.40.0       R.oo_1.26.0               BiocSingular_1.22.0      </span></span>
<span><span class="co">## [139] VariantAnnotation_1.52.0  RCurl_1.98-1.16           magrittr_2.0.3           </span></span>
<span><span class="co">## [142] Formula_1.2-5             GenomeInfoDbData_1.2.13   Rhdf5lib_1.28.0          </span></span>
<span><span class="co">## [145] munsell_0.5.1             Rcpp_1.0.13               viridis_0.6.5            </span></span>
<span><span class="co">## [148] reticulate_1.39.0         stringi_1.8.4             MASS_7.3-61              </span></span>
<span><span class="co">## [151] zlibbioc_1.52.0           plyr_1.8.9                ggstats_0.7.0            </span></span>
<span><span class="co">## [154] parallel_4.4.1            listenv_0.9.1             ggrepel_0.9.6            </span></span>
<span><span class="co">## [157] Biostrings_2.74.0         hms_1.1.3                 circlize_0.4.16          </span></span>
<span><span class="co">## [160] locfit_1.5-9.10           igraph_2.1.1              reshape2_1.4.4           </span></span>
<span><span class="co">## [163] biomaRt_2.62.0            ScaledMatrix_1.14.0       XML_3.99-0.17            </span></span>
<span><span class="co">## [166] evaluate_1.0.1            biovizBase_1.54.0         scran_1.34.0             </span></span>
<span><span class="co">## [169] RcppParallel_5.1.9        BiocManager_1.30.25       tweenr_2.0.3             </span></span>
<span><span class="co">## [172] tzdb_0.4.0                foreach_1.5.2             polyclip_1.10-7          </span></span>
<span><span class="co">## [175] qs_0.27.2                 future_1.34.0             clue_0.3-65              </span></span>
<span><span class="co">## [178] ggforce_0.4.2             rsvd_1.0.5                restfulr_0.0.15          </span></span>
<span><span class="co">## [181] AnnotationFilter_1.30.0   viridisLite_0.4.2         ragg_1.3.3               </span></span>
<span><span class="co">## [184] OrganismDbi_1.48.0        memoise_2.0.1             beeswarm_0.4.0           </span></span>
<span><span class="co">## [187] AnnotationDbi_1.68.0      GenomicAlignments_1.42.0  cluster_2.1.6            </span></span>
<span><span class="co">## [190] timechange_0.3.0          globals_0.16.3</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-chengFlexiplexVersatileDemultiplexer2024" class="csl-entry">
Cheng,O. <em>et al.</em> (2024-03-01) <a href="https://doi.org/10.1093/bioinformatics/btae102" class="external-link">Flexiplex: A
versatile demultiplexer and search tool for omics data</a>.
<em>Bioinformatics</em>, <strong>40</strong>, btae102.
</div>
<div id="ref-liMinimap2PairwiseAlignment2018" class="csl-entry">
Li,H. (2018-09-15) <a href="https://doi.org/10.1093/bioinformatics/bty191" class="external-link">Minimap2: Pairwise
alignment for nucleotide sequences</a>. <em>Bioinformatics</em>,
<strong>34</strong>, 3094â€“3100.
</div>
<div id="ref-liSequenceAlignmentMap2009" class="csl-entry">
Li,H. <em>et al.</em> (2009-08-15) <a href="https://doi.org/10.1093/bioinformatics/btp352" class="external-link">The <span>Sequence
Alignment</span>/<span>Map</span> format and <span>SAMtools</span></a>.
<em>Bioinformatics</em>, <strong>25</strong>, 2078â€“2079.
</div>
<div id="ref-ntranosDiscriminativeLearningApproach2019" class="csl-entry">
Ntranos,V. <em>et al.</em> (2019-02) <a href="https://doi.org/10.1038/s41592-018-0303-9" class="external-link">A discriminative
learning approach to differential expression analysis for single-cell
<span class="nocase">RNA-seq</span></a>. <em>Nat Methods</em>,
<strong>16</strong>, 163â€“166.
</div>
<div id="ref-pengSinglecellRapidCapture2024" class="csl-entry">
Peng,H. <em>et al.</em> (2024-01-30) <a href="https://doi.org/10.1101/2024.01.30.577942" class="external-link">Single-cell <span>Rapid
Capture Hybridization</span> sequencing (<span class="nocase">scRaCH-seq</span>) to reliably detect isoform usage and
coding mutations in targeted genes at a single-cell level</a>.
2024.01.30.577942.
</div>
<div id="ref-rebboahMappingModelingGenomic2021" class="csl-entry">
Rebboah,E. <em>et al.</em> (2021-10-07) <a href="https://doi.org/10.1186/s13059-021-02505-w" class="external-link">Mapping and modeling
the genomic basis of differential <span>RNA</span> isoform expression at
single-cell resolution with <span class="nocase">LR-Split-seq</span></a>. <em>Genome Biology</em>,
<strong>22</strong>, 286.
</div>
<div id="ref-tianComprehensiveCharacterizationSinglecell2021" class="csl-entry">
Tian,L. <em>et al.</em> (2021-11-11) <a href="https://doi.org/10.1186/s13059-021-02525-6" class="external-link">Comprehensive
characterization of single-cell full-length isoforms in human and mouse
with long-read sequencing</a>. <em>Genome Biology</em>,
<strong>22</strong>, 310.
</div>
<div id="ref-WhatMaximumNumber" class="csl-entry">
<a href="https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-" class="external-link">What
is the maximum number of cells that can be profiled?</a>
</div>
<div id="ref-youIdentificationCellBarcodes2023" class="csl-entry">
You,Y. <em>et al.</em> (2023-04-06) <a href="https://doi.org/10.1186/s13059-023-02907-y" class="external-link">Identification of cell
barcodes from long-read single-cell <span class="nocase">RNA-seq</span>
with <span>BLAZE</span></a>. <em>Genome Biology</em>,
<strong>24</strong>, 66.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Changqing Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
