<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES • IgniteRNAseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IgniteRNAseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FLAMESWorkflow.html">Igniting full-length isoform and mutation analysis of single-cell RNA-seq data with FLAMES</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://ChangqingW.github.io/IgniteRNAseq">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Igniting full-length isoform and mutation
analysis of single-cell RNA-seq data with FLAMES</h1>
                        <h4 data-toc-skip class="author">Changqing
Wang</h4>
            <address class="author_afil">
      The Walter and Eliza Hall Institute of Medical Research, 1G Royal
Parade, Parkville, VIC 3052, Melbourne, Australia; Department of Medical
Biology, The University of Melbourne, Parkville, VIC 3010, Melbourne,
Australia<br><h4 data-toc-skip class="date">16 July 2024</h4>
      

      <div class="hidden name"><code>FLAMESWorkflow.Rmd</code></div>

    </address>
</div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>In this vignette, we will walk through the <code>FLAMES</code>
single-cell pipeline, from de-multiplexing raw FASTQ files to
identifying and quantifying transcripts. We then perform analysis
including single nucleotide polymorphism (SNP), differential transcript
expression analysis and differential transcript usage analysis. We will
use the cell line mixture data from Tian <em>et al.</em> (2021) <span class="citation">[@tianComprehensiveCharacterizationSinglecell2021]</span>.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>While the popular single-cell kits allows profiling the transcriptome
of up to 10,000 cells <span class="citation">[@WhatMaximumNumber]</span>
at once, the traditional short-read sequencing (100-200 nts) limits our
ability to unambiguously profile <strong>isoform expression</strong> and
<strong>mutations</strong>. To overcome this issue, we can combine
long-read sequencing with existing single-cell kits.</p>
<p>Though long-read sequencing can provide read lengths several orders
of magnitude larger than the typical short-read sequencing, it comes at
much steeper cost and lower throughput, to accommodate this, we profile
a smaller sub-sample of cells in our <em>FLT-seq</em> protocol<span class="citation">[@tianComprehensiveCharacterizationSinglecell2021]</span>:
<img src="FLT-seq.png" width="400"></p>
<p>There are also many other protocols that combine single-cell with
long-read sequencing, including <em>scRaCH-seq</em> which profiles gene
panel <span class="citation">[@pengSinglecellRapidCapture2024]</span>,
<em>LR-Split-seq</em> which uses combinatorial barcoding and does not
require microfluidic equipments <span class="citation">[@rebboahMappingModelingGenomic2021]</span>.</p>
</div>
<div class="section level2">
<h2 id="set-up">Set-up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/filtered_sorted.bam"</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/filtered_sorted.bam.bai"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"timeout"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12751214/files/filtered_sorted.bam?download=1"</span>, <span class="st">"filtered_sorted.bam"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12751214/files/filtered_sorted.bam.bai?download=1"</span>, <span class="st">"filtered_sorted.bam.bai"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="va">bam</span> <span class="op">&lt;-</span> <span class="st">"filtered_sorted.bam"</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">bam</span> <span class="op">&lt;-</span> <span class="st">"~/filtered_sorted.bam"</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"~/subset_GRCh38.fa"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12751214/files/subset_GRCh38.fa.gz?download=1"</span>, <span class="st">"subset_GRCh38.fa.gz"</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span>  <span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span><span class="st">"subset_GRCh38.fa.gz"</span>, destname <span class="op">=</span> <span class="st">"subset_GRCh38.fa"</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">fa</span> <span class="op">&lt;-</span> <span class="st">"subset_GRCh38.fa"</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">fa</span> <span class="op">&lt;-</span> <span class="st">"~/subset_GRCh38.fa"</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/OliverVoogd/FLAMES" class="external-link">FLAMES</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="flames-pipeline">
<code>FLAMES</code> pipeline<a class="anchor" aria-label="anchor" href="#flames-pipeline"></a>
</h2>
<p><img src="FLAMES_pipeline.png" width="400"></p>
<div class="section level3">
<h3 id="read-pre-processing">Read pre-processing<a class="anchor" aria-label="anchor" href="#read-pre-processing"></a>
</h3>
<p>The <code>FLAMES</code> pipeline starts with de-multiplexing the
FASTQ reads, i.e. identifiying the cell barcode. For protocols that use
the 10x single-cell kits, the raw reads will look like the following,
where the 10x adapter sequences and polyT tails are highlighted. <img src="viewer_mode.png" width="400"><img src="10x_3.png" width="400"></p>
<p><code>FLAMES</code> will locate the cell barcode by locating the
adjacent adapter sequences , and match the barcode against an allow-list
(if it is provided). We can specify the above 10x 3’ protocol as
follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">protocol_pattern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  primer <span class="op">=</span> <span class="st">"CTACACGACGCTCTTCCGATCT"</span>,</span>
<span>  BC <span class="op">=</span> <span class="st">"NNNNNNNNNNNNNNNN"</span>,</span>
<span>  UMI <span class="op">=</span> <span class="st">"NNNNNNNNNNNN"</span>, </span>
<span>  polyT <span class="op">=</span> <span class="st">"TTTTTTTTT"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that the order has to be the same as they appear in the
protocol. The names can be arbitrary for each sequence, but the names
<strong><code>UMI</code> and <code>BC</code> are required</strong> .</p>
<p>Let’s try de-multiplexing a couple reads with
<code><a href="https://rdrr.io/pkg/FLAMES/man/find_barcode.html" class="external-link">FLAMES::find_barcode</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare a barcode allow-list</span></span>
<span><span class="va">bc_allow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_allow.tsv"</span><span class="op">)</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bc_allow.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>, </span>
<span>  destname <span class="op">=</span> <span class="va">bc_allow</span>, remove <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/find_barcode.html" class="external-link">find_barcode</a></span><span class="op">(</span></span>
<span>  <span class="co"># raw fastq file to be de-multiplexed</span></span>
<span>  fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  <span class="co"># outputs</span></span>
<span>  stats_out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_stat"</span><span class="op">)</span>,</span>
<span>  reads_out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>,</span>
<span>  barcodes_file <span class="op">=</span> <span class="va">bc_allow</span>,</span>
<span>  pattern <span class="op">=</span> <span class="va">protocol_pattern</span>, <span class="co"># our protocol pattern</span></span>
<span>  threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## FLEXIPLEX 0.96.2</span></span>
<span><span class="co">## Setting max barcode edit distance to 2</span></span>
<span><span class="co">## Setting max flanking sequence edit distance to 8</span></span>
<span><span class="co">## Setting read IDs to be  replaced</span></span>
<span><span class="co">## Setting number of threads to 1</span></span>
<span><span class="co">## Search pattern: </span></span>
<span><span class="co">## primer: CTACACGACGCTCTTCCGATCT</span></span>
<span><span class="co">## BC: NNNNNNNNNNNNNNNN</span></span>
<span><span class="co">## UMI: NNNNNNNNNNNN</span></span>
<span><span class="co">## polyT: TTTTTTTTT</span></span>
<span><span class="co">## Setting known barcodes from /tmp/Rtmpznvia7/file4e540fb9202/bc_allow.tsv</span></span>
<span><span class="co">## Number of known barcodes: 143</span></span>
<span><span class="co">## Searching for barcodes...</span></span>
<span><span class="co">## Number of reads processed: 393</span></span>
<span><span class="co">## Number of reads where a barcode was found: 368</span></span>
<span><span class="co">## Number of reads where more than one barcode was found: 4</span></span>
<span><span class="co">## All done!</span></span>
<span><span class="co">## Skipping TSO trimming...</span></span></code></pre>
<p><code>FLAMES</code> will find and trim the specified pattern, save
the barcode and UMI to the identifier field of the read, let check the
first raw read:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>           n_max <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "@50409b8e-c5c8-4b86-8d2b-95effbb4434c"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   </span></span>
<span><span class="co">## [2] "GTTGCCGTTCAGTTGCTAAGCAGTGGTATCAACGCGATCCGCTGGGGCCTTGCGCGTTGATATGATTGGTCGGCAGTGAGCGGTCCTCTTTTCCTCCTCTCCCGGCTCGGCGCCGTAGCCATCATGAATGACAGTAACCATCCGGACCAGGAAGTTCATGACCAACCCAAATCTGCTCAGAGGAAACAGATGGTCATTGATATCCTTCATCCTGGGAAGGCAACAGTACCAAAAGGTGAAATTCAGGAAAAGCTGGCCAAAATGTACAAAACCACACCAGATGTCATCTTTGTATTTGGATTCAGAACCCACTTCAATTGGTAGCAAGACCACTAGCTTTGGCATGATCTATGATTCTTTAGATTATGCAAGAAGATGAGCCTAAACACAGACTGGCAAGACATGGCCTTTATGAGAGAAAGACCTCCGAAAACAGCAGAAAAGAGACACAAGAACAGAATGAAGAAGGTCAGGGGCACGGCGAGAGTAATGTTGGTGCTATGCAAAAAGTGAGTGGAGATTGGATACAGCCAAGGAGTAGATCTGCAGTGACTTGATGTTTGCTGTGATGTGCAGATTTCTGAGAGGACAAATAAACTAAAAGCTCCTACTAAAAAAAAAAAAAAAAAAAAAAGACAGCCCGGTACTACTACTGACCCGAGATCAGGAAGAGCGTCGTGTAGAGCAATACGTAAC"</span></span></code></pre>
<p>and the first de-multiplexed read:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>, n_max <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "@CGGGTCAGTAGTAGTA_CCGGGCTGTCTT#50409b8e-c5c8-4b86-8d2b-95effbb4434c_-1of1"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           </span></span>
<span><span class="co">## [2] "AGTAGGAGCTTTTAGTTTATTTGTCCTCTCAGAAATCTGCACATCACAGCAAACATCAAGTCACTGCAGATCTACTCCTTGGCTGTATCCAATCTCCACTCACTTTTTGCATAGCACCAACATTACTCTCGCCGTGCCCCTGACCTTCTTCATTCTGTTCTTGTGTCTCTTTTCTGCTGTTTTCGGAGGTCTTTCTCTCATAAAGGCCATGTCTTGCCAGTCTGTGTTTAGGCTCATCTTCTTGCATAATCTAAAGAATCATAGATCATGCCAAAGCTAGTGGTCTTGCTACCAATTGAAGTGGGTTCTGAATCCAAATACAAAGATGACATCTGGTGTGGTTTTGTACATTTTGGCCAGCTTTTCCTGAATTTCACCTTTTGGTACTGTTGCCTTCCCAGGATGAAGGATATCAATGACCATCTGTTTCCTCTGAGCAGATTTGGGTTGGTCATGAACTTCCTGGTCCGGATGGTTACTGTCATTCATGATGGCTACGGCGCCGAGCCGGGAGAGGAGGAAAAGAGGACCGCTCACTGCCGACCAATCATATCAACGCGCAAGGCCCCAGCGGATCGCGTTGATACCACTGCTTAGCAACTGAACGGCAAC"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>Once the reads are processed, they are ready to be aligned to the
reference genome. <code>FLAMES</code> calls <code>minimap2</code> to
perform alignment and <code>samtoools</code> to sort and index the BAM
file, so we need to make sure the binaries are available in
<code>$PATH</code> for <code>FLAMES</code> to find it.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/sys_which.html" class="external-link">sys_which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/minimap2_align.html" class="external-link">minimap2_align</a></span><span class="op">(</span></span>
<span>        config <span class="op">=</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/SIRV_config_default.json'</span>, package <span class="op">=</span> <span class="st">'FLAMES'</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>        fa_file <span class="op">=</span> <span class="va">fa</span>,</span>
<span>        fq_in <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultiplexed.fq"</span><span class="op">)</span>,</span>
<span>        annot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"filtered.gtf"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span>,</span>
<span>        outdir <span class="op">=</span> <span class="va">outdir</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><code>FLAMES</code> then performs gene quantification, isoform
identification and isoform quantification. For the interest of time we
will skip running them individually, as <code>FLAMES</code> also
provides the <code>sc_long_pipeline</code> wrapper to do all steps.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/sc_long_pipeline.html" class="external-link">sc_long_pipeline</a></span><span class="op">(</span></span>
<span>  genome_fa <span class="op">=</span> <span class="va">fa</span>,</span>
<span>  fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.gtf.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span>
<span>  outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>  barcodes_file <span class="op">=</span> <span class="va">bc_allow</span>,</span>
<span>  minimap2 <span class="op">=</span> <span class="st">'~/k8-1.2/k8-x86_64-Linux'</span>, <span class="co"># apt does not package k8 with minimap2</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h2>
<p>After running the pipeline, we get a
<code>SingleCellExperiment</code> object and we can finally start some
analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce.qs"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 15076 162 </span></span>
<span><span class="co">## metadata(1): OutputFiles</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(15076): ENSG00000000419.14_50934870_50958532_1</span></span>
<span><span class="co">##   ENSG00000001036.14_143495487_143511720_2 ... ENST00000713574.1</span></span>
<span><span class="co">##   ENST00000713579.1</span></span>
<span><span class="co">## rowData names(3): transcript_id gene_id FSM_match</span></span>
<span><span class="co">## colnames(162): AAAGAACAGCGATCGA AACAGGGTCCAGTGCG ... TTTCAGTCACCAGTAT</span></span>
<span><span class="co">##   TTTGGAGCACAAAGTA</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(1): gene</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 14555 162 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(14555): ENSG00000179820.16 ENSG00000114062.22 ... ENSG00000212907.2</span></span>
<span><span class="co">##   ENSG00000228253.1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(162): AAAGAACAGCGATCGA AACAGGGTCCAGTGCG ... TTTCAGTCACCAGTAT</span></span>
<span><span class="co">##   TTTGGAGCACAAAGTA</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>We can use Bioconductor packages for normalization, dimensionality
reduction and clustering:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">cluster_leiden</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/communities.html" class="external-link">membership</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'sizeFactor'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/sce_preprocessing-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="mutation-analysis">Mutation analysis<a class="anchor" aria-label="anchor" href="#mutation-analysis"></a>
</h2>
<div class="section level3">
<h3 id="mutation-discovery">Mutation discovery<a class="anchor" aria-label="anchor" href="#mutation-discovery"></a>
</h3>
<p>Next, we get the bulk allele count to see if there is any potential
SNP mutation in our data.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/find_variants.html" class="external-link">find_variants</a></span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">fa</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"filtered.gtf"</span>, package <span class="op">=</span> <span class="st">"IgniteRNAseq"</span><span class="op">)</span>,</span>
<span>  min_nucleotide_depth <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  homopolymer_window <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  annotated_region_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names_from <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10:12:44 Reading reference ...</span></span></code></pre>
<pre><code><span><span class="co">## 10:12:45 Reading annotation ...</span></span></code></pre>
<pre><code><span><span class="co">## 10:12:45 Got 1 bam file, parallelizing over each region ...</span></span></code></pre>
<pre><code><span><span class="co">## 10:13:37 Merging results ...</span></span></code></pre>
<pre><code><span><span class="co">## 10:13:37 Calculating homopolymer percentages ...</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,088 × 10</span></span></span>
<span><span class="co">##    seqnames   pos nucleotide count    sum     freq ref   bam_path   region homopolymer_pct</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> chrM      <span style="text-decoration: underline;">7</span>021 +           <span style="text-decoration: underline;">1</span>553 <span style="text-decoration: underline;">249</span>499 0.006<span style="text-decoration: underline;">22</span>  T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> chrM      <span style="text-decoration: underline;">7</span>020 -           <span style="text-decoration: underline;">1</span>888 <span style="text-decoration: underline;">249</span>484 0.007<span style="text-decoration: underline;">57</span>  G     filtered_… MT-CO1           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> chrM      <span style="text-decoration: underline;">7</span>020 A           <span style="text-decoration: underline;">1</span>632 <span style="text-decoration: underline;">249</span>484 0.006<span style="text-decoration: underline;">54</span>  G     filtered_… MT-CO1           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> chrM      <span style="text-decoration: underline;">7</span>020 C            117 <span style="text-decoration: underline;">249</span>484 0.000<span style="text-decoration: underline;">469</span> G     filtered_… MT-CO1           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> chrM      <span style="text-decoration: underline;">7</span>020 T            457 <span style="text-decoration: underline;">249</span>484 0.001<span style="text-decoration: underline;">83</span>  G     filtered_… MT-CO1           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> chrM      <span style="text-decoration: underline;">7</span>022 +            565 <span style="text-decoration: underline;">249</span>539 0.002<span style="text-decoration: underline;">26</span>  T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> chrM      <span style="text-decoration: underline;">7</span>021 -           <span style="text-decoration: underline;">2</span>293 <span style="text-decoration: underline;">249</span>499 0.009<span style="text-decoration: underline;">19</span>  T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> chrM      <span style="text-decoration: underline;">7</span>021 A            432 <span style="text-decoration: underline;">249</span>499 0.001<span style="text-decoration: underline;">73</span>  T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> chrM      <span style="text-decoration: underline;">7</span>021 C           <span style="text-decoration: underline;">2</span>509 <span style="text-decoration: underline;">249</span>499 0.010<span style="text-decoration: underline;">1</span>   T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> chrM      <span style="text-decoration: underline;">7</span>021 G           <span style="text-decoration: underline;">1</span>160 <span style="text-decoration: underline;">249</span>499 0.004<span style="text-decoration: underline;">65</span>  T     filtered_… MT-CO1           0.333</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,078 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutation_discovery_tb</span> <span class="op">&lt;-</span> <span class="va">mutation_discovery_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">freq</span> <span class="op">&gt;</span> <span class="fl">0.2</span>, <span class="va">freq</span> <span class="op">&lt;</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mutation_discovery_tb</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">##   seqnames      pos nucleotide  count    sum  freq ref     bam_path region homopolymer_pct</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chrM         <span style="text-decoration: underline;">7</span>028 T          <span style="text-decoration: underline;">130</span>634 <span style="text-decoration: underline;">249</span>993 0.523 C       filtere… MT-CO1           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chrM         <span style="text-decoration: underline;">8</span>251 A          <span style="text-decoration: underline;">104</span>893 <span style="text-decoration: underline;">266</span>526 0.394 G       filtere… MT-CO2           0.5  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr19    48<span style="text-decoration: underline;">965</span>830 C           <span style="text-decoration: underline;">85</span>423 <span style="text-decoration: underline;">178</span>620 0.478 charac… filtere… FTL           -<span style="color: #BB0000;">Inf</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> chr19    48<span style="text-decoration: underline;">965</span>830 T           <span style="text-decoration: underline;">84</span>346 <span style="text-decoration: underline;">178</span>620 0.472 charac… filtere… FTL           -<span style="color: #BB0000;">Inf</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> chrM         <span style="text-decoration: underline;">9</span>123 A           <span style="text-decoration: underline;">66</span>376 <span style="text-decoration: underline;">220</span>334 0.301 G       filtere… MT-AT…           0.333</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> chrM        <span style="text-decoration: underline;">11</span>719 A           <span style="text-decoration: underline;">51</span>284 <span style="text-decoration: underline;">151</span>846 0.338 G       filtere… MT-ND4           0.333</span></span></code></pre>
<p>As we can see, there is a couple of potential SNP mutations, let’s
see if they are correlated with our clustering.</p>
</div>
<div class="section level3">
<h3 id="mutation-calling">Mutation calling<a class="anchor" aria-label="anchor" href="#mutation-calling"></a>
</h3>
<p><code>FLAMES</code> offers the <code>sc_mutations</code> function to
call mutations at the single-cell level:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps_tb</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/sc_mutations.html" class="external-link">sc_mutations</a></span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">mutation_discovery_tb</span><span class="op">$</span><span class="va">seqnames</span><span class="op">)</span>,</span>
<span>  positions <span class="op">=</span> <span class="va">mutation_discovery_tb</span><span class="op">$</span><span class="va">pos</span>,</span>
<span>  indel <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  barcodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10:14:10 Got 1 bam file, parallelizing over each position ...</span></span></code></pre>
<pre><code><span><span class="co">## 10:15:08 Merging results ...</span></span></code></pre>
<p>Lets visualize the SNPs on our previous UMAP:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snps_tb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   allele barcode          allele_count cell_total_reads     pct   pos seqname</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> A      AAAGAACAGCGATCGA            0             <span style="text-decoration: underline;">1</span>030 0        <span style="text-decoration: underline;">7</span>028 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> A      AACAGGGTCCAGTGCG            2             <span style="text-decoration: underline;">1</span>649 0.001<span style="text-decoration: underline;">21</span>  <span style="text-decoration: underline;">7</span>028 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> A      AAGTCGTCAGTAGGAC            9             <span style="text-decoration: underline;">2</span>299 0.003<span style="text-decoration: underline;">91</span>  <span style="text-decoration: underline;">7</span>028 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> A      AATCACGGTCCTCCTA            7             <span style="text-decoration: underline;">1</span>728 0.004<span style="text-decoration: underline;">05</span>  <span style="text-decoration: underline;">7</span>028 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> A      AATCGACAGGGTGGGA            5             <span style="text-decoration: underline;">1</span>611 0.003<span style="text-decoration: underline;">10</span>  <span style="text-decoration: underline;">7</span>028 chrM   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> A      AATCGACTCGTTTACT            2             <span style="text-decoration: underline;">1</span>196 0.001<span style="text-decoration: underline;">67</span>  <span style="text-decoration: underline;">7</span>028 chrM</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr19</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chr19'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">48965830</span><span class="op">)</span></span>
<span><span class="va">chr19</span> <span class="op">&lt;-</span> <span class="va">chr19</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chr19</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chr19</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-1.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_7028</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">7028</span><span class="op">)</span></span>
<span><span class="va">chrM_7028</span> <span class="op">&lt;-</span> <span class="va">chrM_7028</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_7028</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_7028</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-2.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_8251</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">8251</span><span class="op">)</span></span>
<span><span class="va">chrM_8251</span> <span class="op">&lt;-</span> <span class="va">chrM_8251</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_8251</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_8251</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-3.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_9123</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">9123</span><span class="op">)</span></span>
<span><span class="va">chrM_9123</span> <span class="op">&lt;-</span> <span class="va">chrM_9123</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_9123</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_9123</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-4.png" width="700"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrM_11719</span> <span class="op">&lt;-</span> <span class="va">snps_tb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele_count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">pct</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqname</span> <span class="op">==</span> <span class="st">'chrM'</span>, <span class="va">pos</span> <span class="op">==</span> <span class="fl">11719</span><span class="op">)</span></span>
<span><span class="va">chrM_11719</span> <span class="op">&lt;-</span> <span class="va">chrM_11719</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">chrM_11719</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>,<span class="st">"allele"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="va">chrM_11719</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/view_mutations-5.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-transcript-experssion-analysis">Differential Transcript Experssion analysis<a class="anchor" aria-label="anchor" href="#differential-transcript-experssion-analysis"></a>
</h2>
<p>Since we have the transcript count matrix from <code>FLAMES</code>,
we can do some DTE analysis:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span>, <span class="va">colSums</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">identity</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="va">colSums</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">identity</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">pseudo_bulk</span><span class="op">[</span>, <span class="op">-</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="va">dgelist</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dgelist</span><span class="op">)</span>, , <span class="co"># bias?</span></span>
<span>  keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="va">dgelist</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/DTE-1.png" width="700"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dgelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dgelist</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span>
<span><span class="co"># means model</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dgelist</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span>, data <span class="op">=</span> <span class="va">dgelist</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*\\$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" .*$"</span>, <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span></span>
<span><span class="va">dge_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dgelist</span>, <span class="va">design</span>, save.plot <span class="op">=</span> <span class="cn">T</span>, plot <span class="op">=</span> <span class="cn">F</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">dge_v</span>, <span class="va">dge_v</span><span class="op">$</span><span class="va">design</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span></span>
<span>    MalevsFemale <span class="op">=</span> <span class="st">"groupA - groupB"</span>,</span>
<span>    levels <span class="op">=</span> <span class="va">dge_v</span><span class="op">$</span><span class="va">design</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    logFC AveExpr      t  P.Value adj.P.Val    B</span></span>
<span><span class="co">## ENST00000621160.5  -7.20    7.01 -11.71 1.29e-06    0.0103 5.15</span></span>
<span><span class="co">## ENST00000284440.9   5.95    6.64  11.00 2.15e-06    0.0103 4.92</span></span>
<span><span class="co">## ENST00000618200.4  -7.91    6.81 -11.03 2.10e-06    0.0103 4.46</span></span>
<span><span class="co">## ENST00000649529.1  -5.55    7.65  -9.11 9.85e-06    0.0289 3.80</span></span>
<span><span class="co">## ENST00000316403.15  5.12    5.99   8.87 1.21e-05    0.0289 3.58</span></span>
<span><span class="co">## ENST00000412585.7  -4.57    6.00  -8.60 1.55e-05    0.0306 3.37</span></span>
<span><span class="co">## ENST00000361157.11 -4.61    6.96  -8.41 1.85e-05    0.0306 3.27</span></span>
<span><span class="co">## ENST00000361566.7  -5.38    8.03  -8.29 2.06e-05    0.0306 3.19</span></span>
<span><span class="co">## ENST00000378292.9   6.48    5.22   9.03 1.05e-05    0.0289 3.18</span></span>
<span><span class="co">## ENST00000368716.9   4.90    6.44   8.25 2.14e-05    0.0306 3.14</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">]</span>, <span class="st">'gene'</span><span class="op">)</span>, <span class="st">'UMAP'</span>, </span>
<span>                       colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>raw_expr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">'ENST00000621160.5'</span>, <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">!=</span> <span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMESWorkflow_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="differential-transcript-usage-analysis">Differential Transcript Usage analysis<a class="anchor" aria-label="anchor" href="#differential-transcript-usage-analysis"></a>
</h2>
<p>WIP</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">OutputFiles</span><span class="op">$</span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://zenodo.org/records/12751214/files/isoform_FSM_annotation.csv?download=1'</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">sce</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">OutputFiles</span><span class="op">$</span><span class="va">outdir</span>, <span class="st">"isoform_FSM_annotation.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">dtu_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FLAMES/man/sc_DTU_analysis.html" class="external-link">sc_DTU_analysis</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading isoform_FSM_annotation.csv ...</span></span>
<span><span class="co">## Selecting transcript_ids with full splice match ...</span></span>
<span><span class="co">## Summing transcripts with same FSM ... </span></span>
<span><span class="co">## Creating FSM_count.csv.gz ...</span></span>
<span><span class="co">## /tmp/Rtmpznvia7/FSM_count.csv.gz saved.</span></span>
<span><span class="co">##   12340 FSM_match(s) found.</span></span>
<span><span class="co">## Filtering for genes with at least 2 detected isforms ...      8895 FSM_match(s) left.</span></span>
<span><span class="co">## Keeping only the top 4 expressed FSM_matches for each gene ...      7783 FSM_match(s) left.</span></span>
<span><span class="co">## Aggregating counts by cluster labels ...</span></span>
<span><span class="co">## Filtering isoforms ... 7783  transcript_id(s) remaining.</span></span>
<span><span class="co">## Performing Chi-square tests ...</span></span>
<span><span class="co">##   |                                                                                        |                                                                                |   0%  |                                                                                        |                                                                                |   1%  |                                                                                        |=                                                                               |   1%  |                                                                                        |=                                                                               |   2%  |                                                                                        |==                                                                              |   2%  |                                                                                        |==                                                                              |   3%  |                                                                                        |===                                                                             |   3%  |                                                                                        |===                                                                             |   4%  |                                                                                        |====                                                                            |   4%  |                                                                                        |====                                                                            |   5%  |                                                                                        |====                                                                            |   6%  |                                                                                        |=====                                                                           |   6%  |                                                                                        |=====                                                                           |   7%  |                                                                                        |======                                                                          |   7%  |                                                                                        |======                                                                          |   8%  |                                                                                        |=======                                                                         |   8%  |                                                                                        |=======                                                                         |   9%  |                                                                                        |========                                                                        |   9%  |                                                                                        |========                                                                        |  10%  |                                                                                        |========                                                                        |  11%  |                                                                                        |=========                                                                       |  11%  |                                                                                        |=========                                                                       |  12%  |                                                                                        |==========                                                                      |  12%  |                                                                                        |==========                                                                      |  13%  |                                                                                        |===========                                                                     |  13%  |                                                                                        |===========                                                                     |  14%  |                                                                                        |============                                                                    |  14%  |                                                                                        |============                                                                    |  15%  |                                                                                        |============                                                                    |  16%  |                                                                                        |=============                                                                   |  16%  |                                                                                        |=============                                                                   |  17%  |                                                                                        |==============                                                                  |  17%  |                                                                                        |==============                                                                  |  18%  |                                                                                        |===============                                                                 |  18%  |                                                                                        |===============                                                                 |  19%  |                                                                                        |================                                                                |  19%  |                                                                                        |================                                                                |  20%  |                                                                                        |================                                                                |  21%  |                                                                                        |=================                                                               |  21%  |                                                                                        |=================                                                               |  22%  |                                                                                        |==================                                                              |  22%  |                                                                                        |==================                                                              |  23%  |                                                                                        |===================                                                             |  23%  |                                                                                        |===================                                                             |  24%  |                                                                                        |====================                                                            |  24%  |                                                                                        |====================                                                            |  25%  |                                                                                        |====================                                                            |  26%  |                                                                                        |=====================                                                           |  26%  |                                                                                        |=====================                                                           |  27%  |                                                                                        |======================                                                          |  27%  |                                                                                        |======================                                                          |  28%  |                                                                                        |=======================                                                         |  28%  |                                                                                        |=======================                                                         |  29%  |                                                                                        |========================                                                        |  29%  |                                                                                        |========================                                                        |  30%  |                                                                                        |========================                                                        |  31%  |                                                                                        |=========================                                                       |  31%  |                                                                                        |=========================                                                       |  32%  |                                                                                        |==========================                                                      |  32%  |                                                                                        |==========================                                                      |  33%  |                                                                                        |===========================                                                     |  33%  |                                                                                        |===========================                                                     |  34%  |                                                                                        |============================                                                    |  34%  |                                                                                        |============================                                                    |  35%  |                                                                                        |============================                                                    |  36%  |                                                                                        |=============================                                                   |  36%  |                                                                                        |=============================                                                   |  37%  |                                                                                        |==============================                                                  |  37%  |                                                                                        |==============================                                                  |  38%  |                                                                                        |===============================                                                 |  38%  |                                                                                        |===============================                                                 |  39%  |                                                                                        |================================                                                |  39%  |                                                                                        |================================                                                |  40%  |                                                                                        |================================                                                |  41%  |                                                                                        |=================================                                               |  41%  |                                                                                        |=================================                                               |  42%  |                                                                                        |==================================                                              |  42%  |                                                                                        |==================================                                              |  43%  |                                                                                        |===================================                                             |  43%  |                                                                                        |===================================                                             |  44%  |                                                                                        |====================================                                            |  44%  |                                                                                        |====================================                                            |  45%  |                                                                                        |====================================                                            |  46%  |                                                                                        |=====================================                                           |  46%  |                                                                                        |=====================================                                           |  47%  |                                                                                        |======================================                                          |  47%  |                                                                                        |======================================                                          |  48%  |                                                                                        |=======================================                                         |  48%  |                                                                                        |=======================================                                         |  49%  |                                                                                        |========================================                                        |  49%  |                                                                                        |========================================                                        |  50%  |                                                                                        |========================================                                        |  51%  |                                                                                        |=========================================                                       |  51%  |                                                                                        |=========================================                                       |  52%  |                                                                                        |==========================================                                      |  52%  |                                                                                        |==========================================                                      |  53%  |                                                                                        |===========================================                                     |  53%  |                                                                                        |===========================================                                     |  54%  |                                                                                        |============================================                                    |  54%  |                                                                                        |============================================                                    |  55%  |                                                                                        |============================================                                    |  56%  |                                                                                        |=============================================                                   |  56%  |                                                                                        |=============================================                                   |  57%  |                                                                                        |==============================================                                  |  57%  |                                                                                        |==============================================                                  |  58%  |                                                                                        |===============================================                                 |  58%  |                                                                                        |===============================================                                 |  59%  |                                                                                        |================================================                                |  59%  |                                                                                        |================================================                                |  60%  |                                                                                        |================================================                                |  61%  |                                                                                        |=================================================                               |  61%  |                                                                                        |=================================================                               |  62%  |                                                                                        |==================================================                              |  62%  |                                                                                        |==================================================                              |  63%  |                                                                                        |===================================================                             |  63%  |                                                                                        |===================================================                             |  64%  |                                                                                        |====================================================                            |  64%  |                                                                                        |====================================================                            |  65%  |                                                                                        |====================================================                            |  66%  |                                                                                        |=====================================================                           |  66%  |                                                                                        |=====================================================                           |  67%  |                                                                                        |======================================================                          |  67%  |                                                                                        |======================================================                          |  68%  |                                                                                        |=======================================================                         |  68%  |                                                                                        |=======================================================                         |  69%  |                                                                                        |========================================================                        |  69%  |                                                                                        |========================================================                        |  70%  |                                                                                        |========================================================                        |  71%  |                                                                                        |=========================================================                       |  71%  |                                                                                        |=========================================================                       |  72%  |                                                                                        |==========================================================                      |  72%  |                                                                                        |==========================================================                      |  73%  |                                                                                        |===========================================================                     |  73%  |                                                                                        |===========================================================                     |  74%  |                                                                                        |============================================================                    |  74%  |                                                                                        |============================================================                    |  75%  |                                                                                        |============================================================                    |  76%  |                                                                                        |=============================================================                   |  76%  |                                                                                        |=============================================================                   |  77%  |                                                                                        |==============================================================                  |  77%  |                                                                                        |==============================================================                  |  78%  |                                                                                        |===============================================================                 |  78%  |                                                                                        |===============================================================                 |  79%  |                                                                                        |================================================================                |  79%  |                                                                                        |================================================================                |  80%  |                                                                                        |================================================================                |  81%  |                                                                                        |=================================================================               |  81%  |                                                                                        |=================================================================               |  82%  |                                                                                        |==================================================================              |  82%  |                                                                                        |==================================================================              |  83%  |                                                                                        |===================================================================             |  83%  |                                                                                        |===================================================================             |  84%  |                                                                                        |====================================================================            |  84%  |                                                                                        |====================================================================            |  85%  |                                                                                        |====================================================================            |  86%  |                                                                                        |=====================================================================           |  86%  |                                                                                        |=====================================================================           |  87%  |                                                                                        |======================================================================          |  87%  |                                                                                        |======================================================================          |  88%  |                                                                                        |=======================================================================         |  88%  |                                                                                        |=======================================================================         |  89%  |                                                                                        |========================================================================        |  89%  |                                                                                        |========================================================================        |  90%  |                                                                                        |========================================================================        |  91%  |                                                                                        |=========================================================================       |  91%  |                                                                                        |=========================================================================       |  92%  |                                                                                        |==========================================================================      |  92%  |                                                                                        |==========================================================================      |  93%  |                                                                                        |===========================================================================     |  93%  |                                                                                        |===========================================================================     |  94%  |                                                                                        |============================================================================    |  94%  |                                                                                        |============================================================================    |  95%  |                                                                                        |============================================================================    |  96%  |                                                                                        |=============================================================================   |  96%  |                                                                                        |=============================================================================   |  97%  |                                                                                        |==============================================================================  |  97%  |                                                                                        |==============================================================================  |  98%  |                                                                                        |=============================================================================== |  98%  |                                                                                        |=============================================================================== |  99%  |                                                                                        |================================================================================|  99%  |                                                                                        |================================================================================| 100%</span></span></code></pre>
<pre><code><span><span class="co">## Warning in sc_DTU_analysis(sce): Chi-squared approximation(s) may be incorrect</span></span></code></pre>
<pre><code><span><span class="co">## Results saved to  /tmp/Rtmpznvia7/sc_DTU_analysis.csv</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Changqing Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
